src/identity/caclient.rs

```diff
@@ -99,35 +99,91 @@ impl CertificateProvider for CaClient {
 }
 
 pub mod mock {
+    use std::sync::Arc;
     use std::time::Duration;
 
     use async_trait::async_trait;
+    use tokio::sync::RwLock;
+    use tokio::time::Instant;
 
     use crate::identity::{CertificateProvider, Identity};
-    use crate::tls::{generate_test_certs, Certs};
+    use crate::tls::{generate_test_certs_at, Certs};
 
     use super::*;
 
-    #[derive(Clone, Debug)]
-    pub struct CaClient {
+    #[derive(Default)]
+    struct ClientState {
+        fetches: Vec<Identity>,
+    }
+
+    #[derive(Clone)]
+    pub struct ClientConfig {
         pub cert_lifetime: Duration,
+        pub time_conv: crate::time::Converter,
+        // If non-zero, causes fetch_certificate calls to sleep for the specified duration before
+        // returning. This is helpful to let tests that pause tokio time get more control over code
+        // execution.
+        pub fetch_latency: Duration,
     }
 
-    #[async_trait]
-    impl CertificateProvider for CaClient {
-        async fn fetch_certificate(&self, id: &Identity) -> Result<Certs, Error> {
-            let certs = generate_test_certs(
-                &id.clone().into(),
-                Duration::from_secs(0),
-                self.cert_lifetime,
-            );
-            return Ok(certs);
+    impl Default for ClientConfig {
+        fn default() -> Self {
+            Self {
+                fetch_latency: Duration::ZERO,
+                cert_lifetime: Duration::from_secs(10),
+                time_conv: crate::time::Converter::new(),
+            }
         }
     }
 
+    #[derive(Clone)]
+    pub struct CaClient {
+        cfg: ClientConfig,
+        state: Arc<RwLock<ClientState>>,
```

Comment:
What's the benefit of having a state struct holding the `fetches` field instead of defining it here with type `Arc<RwLock<Vec<Identity>>>`?

Reply:
At some point I was thinking of putting some more fields there but then abandoned the idea. Can replace ClientState with Vec<Identity> if anybody cares.